BUG: align with broadcast_axis, #13194

pfrcks opened the pull request:
Closes #13194 
- Implemented the _easy fix_
- Future work : To implement broadcast axis remaining functionality.
- All tests in test_generic.py passed.

pfrcks committed:
BUG: GH13194 fixed

jreback commented:
tests

we r on 0.18.2

pfrcks commented:
Hi,

I seem to have made the documentation in the wrong file. I will correct it and push the changes.
Rest all is OK, right?
This is my first pull request to pandas, sorry for the noob errors.

pfrcks committed:
DOC : Change in release notes

pfrcks commented:
@jreback I have corrected the documentation to the proper version file. Is this everything?
I have ran the tests also.
Am I missing something?

jreback commented:
you need to add tests which fail before this patch and pass after

kdebrab commented:
Apart from the indispensable unit test, I'd propose to move [the lines](https://github.com/pydata/pandas/blob/master/pandas/core/generic.py#L4172):
```python
if axis is not None:
    axis = self._get_axis_number(axis)
```
to right before the `if broadcast_axis == 1` block, such that broadcasting will work also in case align is called with `axis='index'` instead of `axis=0`.

Thanks!

pfrcks committed:
TST: Added tests for #13194

pfrcks committed:
CLN: Changed location of axis check

pfrcks commented:
@jreback @kdebrab I've included the unit test and have also moved the location of the axis check.
Can you please take a look?

jreback commented on pandas/tests/test_generic.py:
```diff
@@ -1874,6 +1874,52 @@ def test_pipe_panel(self):
             result = wp.pipe((f, 'y'), x=1, y=1)
 
+    def test_align_joins(self):
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
```
add the issue number as a comment here

rename to ``test_align_broadcast_axis``

jreback commented on pandas/tests/test_generic.py:
```diff
@@ -1874,6 +1874,52 @@ def test_pipe_panel(self):
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
+        ts = Series([5., 6., 7.])
+        check = df.align(ts, join='outer', axis=0, broadcast_axis=1)
+        df1 = DataFrame(check[0])
```
all of these tests need to go in ``tests/frame/test_axis_select_reinde`` and/or ``tests/series/test_indexing``. right after the other align tests.

jreback replied:
model them after the existing tests.

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. (:issue:`13194`) 
```
be more descriptive of what you fixed here. remember a user is reading this.

pfrcks committed:
TST : Renamed and changed the location of unittests.

pfrcks commented:
@jreback Have made the required changes. 
Please comment if something's lacking.

pfrcks committed:
DOC : improved the documentation in release notes

sinhrks commented on pandas/tests/frame/test_axis_select_reindex.py:
```diff
@@ -880,3 +880,53 @@ def test_reindex_multi(self):
+
+        # For 'inner' join
+        result = df.align(ts, join='inner', axis=0, broadcast_axis=1)
+        result1 = DataFrame(result[0])
```
Not needed. Test returned obj (`result[0]` and `[1]` as it is.

pfrcks replied:
@sinhrks I'm not sure that I understand. Can you please elaborate?

sinhrks replied:
Pls remove `DataFrame(result..)` like below.

```
        result = df.align(...)
        expected1 = ...
        expected2 = ...
        assert_frame_equal(result[0], expected1)
        assert_frame_equal(result[1], expected2)
```

sinhrks commented:
Can u add tests with:
- alternately different `Index` to clheck left and right join.
- `Series.align(DataFrame)`

pfrcks commented:
@sinhrks Sure. It's late night here right now. I will do this in the morning.

pfrcks committed:
TST: Added and modified unittests

pfrcks commented:
@sinhrks I've made some changes. Please take a look and comment. This is what you meant right?

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. Earlier, upon supplying value of join argument as any of the four('outer', 'inner', 'left', 'right' ), align() was giving erraneous output. Align with broadcast_axis specified was using 'inner' join consistently irrespective of the value of 'join' provided when aligning dataframe and s ...
```
Single line of comments pls. these are read by the users and don't need this kind of explanation.

jreback commented on pandas/tests/frame/test_axis_select_reindex.py:
```diff
@@ -880,3 +880,66 @@ def test_reindex_multi(self):
+        assert_frame_equal(result[0], expected1)
+        assert_frame_equal(result[1], expected2)
+
+        # Series.align(DataFrame) tests, 'outer' join
```
these 2 series-frame joins should be in ``tests/series/test_indexing.py``

pfrcks committed:
TST: Restructured and moved unittests

pfrcks commented:
@jreback Done. Kindly check and comment.

jreback commented on pandas/tests/frame/test_axis_select_reindex.py:
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
+        ts = Series([5., 6., 7.])
+
+        result = df.align(ts, join='right', axis=0, broadcast_axis=1)
```
try one of these with ``broadcast_axis='columns'`` and one with ``broadcast_axis='index'`` as well

jreback replied:
``left, right = df.align(....)`` and same below

jreback commented on pandas/tests/frame/test_axis_select_reindex.py:
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
+        ts = Series([5., 6., 7.])
+
+        result = df.align(ts, join='right', axis=0, broadcast_axis=1)
+        expected1 = DataFrame(np.array([[1., 2.], [3., 4.],
```
call these ``left`` and ``right`` (and same below)

jreback commented on pandas/tests/series/test_indexing.py:
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
+        # Series.align(DataFrame) tests, 'outer' join
+        df = DataFrame(np.array([[1., 2.], [3., 4.]]), columns=list('AB'))
+        ts = Series([5., 6., 7.])
+        result = ts.align(df, join='outer', axis=0, broadcast_axis=1)
```
same name scheme here

jreback commented on pandas/tests/series/test_indexing.py:
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
+        expected2 = DataFrame(np.array([[1., 2.], [3., 4.],
+                                        [pd.np.nan, pd.np.nan]]),
+                              columns=list('AB'))
+        self.assertTrue(result[0].equals(expected1))
```
``tm.assert_frame_equal``

jreback commented on doc/source/whatsnew/v0.18.2.txt:
```diff
@@ -149,3 +149,4 @@ Bug Fixes
 - Bug in ``NaT`` - ``Period`` raises ``AttributeError`` (:issue:`13071`)
 - Bug in ``Period`` addition raises ``TypeError`` if ``Period`` is on right hand side (:issue:`13069`)
 - Bug in ``pd.set_eng_float_format()`` that would prevent NaN's from formatting (:issue:`11981`)
+- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with broadcast_axis. (:issue:`13194`) 
```
``Bug in ``

use double-backticks around ``broadcast_axis``.

say ``incorrect output`` (not wrong output)

jreback replied:
say that this was not taking into account the ``how`` parameter when ``broadcast_axis`` was passed

jreback commented on pandas/tests/frame/test_axis_select_reindex.py:
```diff
@@ -880,3 +880,47 @@ def test_reindex_multi(self):
+
+        # For 'right' join on different index
+        result = df.align(ts, join='right', axis=1, broadcast_axis=1)
+        expected1 = DataFrame(np.array([[1., 2.], [3., 4.]]),
```
you don't need to pass these as numpy arrays, just list-of-lists is fine

pfrcks commented:
@jreback 

        right = df.align(ts, join='right', axis=1, broadcast_axis='index')
        expected1_right = DataFrame([[pd.np.nan, pd.np.nan, pd.np.nan],
                                     [pd.np.nan, pd.np.nan, pd.np,nan]],
                                    columns=[0, 1, 2])
        expected2_right = Series([5., 6., 7.])
        assert_frame_equal(right[0], expected1_right)
        assert_series_equal(right[1], expected2_right)  

While trying to run the above code, nosetests gives the error 

AssertionError: Series.index are different

        Series.index classes are not equivalent
        [left]:  Index([u'A', u'B'], dtype='object')
        [right]: RangeIndex(start=0, stop=2, step=1)

However, upon running in the interpreter

>>> right = df.align(ts, join='right', axis=1, broadcast_axis='index')
>>> right[1]
0    5.0
1    6.0
2    7.0
dtype: float64
>>> right[1].index
RangeIndex(start=0, stop=3, step=1)
>>> expected2_right = Series([5., 6., 7.])
>>> expected2_right.index
RangeIndex(start=0, stop=3, step=1)


We find that both indexes are same.
I am currently in pandas_dev environment.
Is this something assert_series_frame is doing?

sinhrks commented on pandas/tests/series/test_indexing.py:
```diff
@@ -1838,3 +1838,26 @@ def test_multilevel_preserve_name(self):
+        self.assertTrue(result[1].equals(expected2))
+
+        # Series.align(DataFrame) tests, 'inner' join
+        result = ts.align(df, join='inner', axis=0, broadcast_axis=1)
```
we can make all tests more comprehensive as "inner" and "outer" outputs the same, and "left" and "right" outputs inversed results.

```
result = df.align(series, ...)
self.assert_frame_equal(result[0], expected1)
self.assert_frame_equal(result[1], expected2)
result = series.align(df, ...)
self.assert_frame_equal(result[0], expected2)
self.assert_frame_equal(result[1], expected1)
```

sinhrks replied:
missed @jreback suggestion for variable names:

```
left, right = df.align(series, ...)
self.assert_frame_equal(left, expected_left)
self.assert_frame_equal(right, expected_right)
left, right = series.align(df, ...)
self.assert_frame_equal(right, expected_left)
self.assert_frame_equal(left, expected_right)
```

pfrcks replied:
@jreback Can you please look into the issue I have reported above.

jreback replied:
a single Ping is enough then you need to wait for comments

jreback commented:
can you rebase / update?

pfrcks commented:
@jreback I'm not sure I understand. I have rebased my_branch with master.

jreback commented:
doesn't look like it
you need to push

pfrcks commented:
@jreback 
I have used the following commands:

`git checkout my_branch`
`git rebase master`
`git push -u origin my_branch`

pfrcks commented:
Sorry I closed my mistake

jreback commented:
got rebase origin/master

jreback commented:
got -> git

jreback commented:
you are rebasing to a local master which is not updated

pfrcks committed:
Cleaned up doc

pfrcks commented:
@jreback 

I followed the steps:

1) `git remote add ups https://github.com/pydata/pandas.git`
2) `git fetch ups`
3) `git checkout master`
4) `git rebase ups/master`

> First, rewinding head to replay your work on top of it...
Applying: BUG: GH13194 fixed
Applying: DOC : Change in release notes
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.18.2.txt
<stdin>:19: trailing whitespace.
- Fixed the bug in ``DataFrame.align()`` which was giving wrong output when supplied with the ``join`` argument. (:issue:`13194`) 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): doc/source/whatsnew/v0.18.2.txt deleted in HEAD and modified in DOC : Change in release notes. Version DOC : Change in release notes of doc/source/whatsnew/v0.18.2.txt left in tree.
Failed to merge in the changes.
Patch failed at 0002 DOC : Change in release notes
The copy of the patch that failed is found in:
   /home/pfrcks/issues/pandas-pfrcks/.git/rebase-apply/patch
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

I removed the whitespace error and then synced the changes but still I am getting the same issue.

jreback commented:
closing, but if you want to update / rebase. pls reopen
